File: java/secretmessage.java
import java.util.Arrays;
import java.util.Scanner;

public class secretmessage {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int N = scan.nextInt();
		while (N-- > 0) {
			String message = scan.next();
			int K = (int) Math.ceil(Math.sqrt(message.length()));
			char[][] map = new char[K][K];
			for (char[] c : map)
				Arrays.fill(c, '*');
			int count = 0;
			for (int i = 0; i < K; i++)
				for (int j = 0; j < K; j++)
					if (count < message.length())
						map[i][j] = message.charAt(count++);
					else
						break;
			StringBuilder sb = new StringBuilder();
			for (int j = 0; j < K; j++)
				for (int i = K - 1; i >= 0; i--)
					if (map[i][j] != '*')
						sb.append(map[i][j]);
			System.out.println(sb);
		}
		scan.close();
	}
}

File: ts/src/secretmessage.ts
import readline = require('readline')

function encrypt (message: string): string {
  const matrix = fillInMatrix(message)
  return rotateMatrix(matrix)
    .reduce((acc, row) => acc.concat(row.filter((c) => c !== '*')), [])
    .join('')
}

function fillInMatrix (message: string): string[][] {
  const matrixLength = Math.ceil(Math.sqrt(message.length))
  const matrix: string[][] = []
  for (let i = 0; i < message.length; i += matrixLength) {
    const row = message.slice(i, i + matrixLength).split('')
    if (row.length < matrixLength) {
      row.push(...Array(matrixLength - row.length).fill('*'))
    }
    matrix.push(row)
  }
  return matrix
}

function rotateMatrix (matrix: string[][]): string[][] {
  matrix.reverse()

  const matrixLength = matrix.length
  const rotatedMatrix: string[][] = []
  for (let i = 0; i < matrixLength; i++) {
    rotatedMatrix.push(matrix.map((row) => row[i] ?? ''))
  }
  return rotatedMatrix
}

let lineCount = 0

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (lineCount > 0) {
      console.log(encrypt(line))
    }
    lineCount++
  })
  .on('close', () => {
    process.exit(0)
  })
